src/dovecot/dovecot-279/dovecot-pigeonhole/doc/man/sieve-filter.1.in
.\" Copyright (c) 2010-2012 Pigeonhole authors, see the included COPYING file
.TH "SIEVE\-FILTER" 1 "2012-01-07" "Pigeonhole for Dovecot v2.0" "Pigeonhole"
.SH NAME
sieve\-filter \- Pigeonhole\(aqs Sieve mailbox filter tool

.PP
\fBWARNING: \fRThis tool is not finished and should \fB*NOT*\fR be used, unless
you feel like testing newly developed features! The behavior described in this
manual page represents the design and not necessarily what the tool currently
implements.
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.B sieve\-filter
.RI [ options ]
.I script\-file
.I source\-mailbox
.RI [ source\-action ]
.SH DESCRIPTION
.PP
The \fBsieve\-filter\fP command is part of the Pigeonhole Project
(\fBpigeonhole\fR(7)), which adds Sieve (RFC 5228) support to the Dovecot
secure IMAP and POP3 server (\fBdovecot\fR(1)).
.PP
The Sieve language was originally meant for filtering messages upon delivery. 
However, there are occasions when it is desirable to filter messages that are
already stored in a mailbox, for instance when a bug in a Sieve script caused
many messages to be delivered incorrectly. Using the sieve\-filter tool it is
possible to apply a Sieve script on all messages in a particular mailbox, making
it possible to delete messages, to store them in a different mailbox and to
change the assigned IMAP flags and keywords. Attempts to send messages to the
outside world are ignored by default for obvious reasons, but, using the proper
command line options, it is possible to capture outgoing mail as well. 
.PP
If no options are specified, the sieve\-filter command runs in a simulation mode
in which it only prints what would be performed, without actually doing
anything. Use the \fB\-e\fP option to activate true script execution. Also, the
source mailbox is opened read\-only by default, so that the source mailbox 
remains unchanged. Use the \fB\-W\fP option to allow changes in the source mailbox.
And even with the \fB\-W\fP option enabled, messages in the source mailbox are
only potentially modified and not (re)moved, unless a \fIsource\-action\fP
argument other than \fBkeep\fP is specified.
.SS CAUTION
Although this is a very useful tool, it can also be very destructive when used
improperly. A small bug in your Sieve script in combination with the wrong
command line options could cause it to discard the wrong e\-mails. And, even if
the source mailbox is opened in read\-only mode to prevent such mishaps, it can
still litter other mailboxes with spurious copies of your e\-mails if your Sieve
script decides to do so. Therefore, users are advised to read this manual
carefully and to use the simulation mode first to check what the script will do.
And, of course:
.PP
\fBMAKING A BACKUP IS IMPERATIVE FOR ANY IMPORTANT MAIL!\fP

.\"------------------------------------------------------------------------
.SH OPTIONS
.TP
.BI \-c\  config\-file
Alternative Dovecot configuration file path.
.TP
.B \-C
Force compilation. By default, the compiled binary is stored on disk. When this
binary is found during the next execution of \fBsieve\-filter\fP and its
modification time is more recent than the script file, it is used and the script
is not compiled again. This option forces the script to be compiled, thus
ignoring any present binary. Refer to \fBsievec\fP(1) for more information about
Sieve compilation.
.TP
.B \-D
Enable Sieve debugging.
.TP
.B \-e
Turns on execution mode. By default, the sieve\-filter command runs in
simulation mode in which it changes nothing, meaning that no mailbox is altered
in any way and no actions are performed. It only prints what would be done.
Using this option the sieve\-filter command becomes active and performs the 
requested actions.
.TP
.BI \-m\  default\-mailbox
The mailbox where the (implicit) \fBkeep\fP Sieve action stores messages. This is
equal to the source mailbox by default. 
.TP
.B \-M
Enable move mode. This will cause all messages that were succesfully stored
somewhere else to be expunged from the source mailbox, regardless of what the 
\fIsource\-action\fP is (refer to Arguments section below). However, if the
Sieve filter decides to keep the message in the source mailbox, it is left
there and not affected by this option. 
.IP
Note that some Sieve actions, such as \fBredirect\fP, don't store the message
anywhere and are thus \- with respect to the fate of the message in the source
mailbox \- treated as if a \fBdiscard\fP action were executed. 
.TP
.BI \-q\  output\-mailbox\  \fR[not\ implemented\ yet]\fP
Store outgoing e\-mail into the indicated \fIoutput\-mailbox\fP. By default,
the sieve\-filter command ignores Sieve actions such as redirect, reject,
vacation and notify, but using this option outgoing messages can be appended to
the indicated mailbox. This option has no effect in simulation mode. Flags of
redirected messages are not preserved. 
.TP
.BI \-Q\  mail\-command\  \fR[not\ implemented\ yet]\fP
Send outgoing e\-mail (e.g. as produced by redirect, reject and vacation)
through the specified program. By default, the sieve\-filter command ignores
Sieve actions such as redirect, reject, vacation and notify, but using this
option outgoing messages can be fed to the \fBstdin\fP of an external shell
command. This option has no effect in simulation mode. Unless you really know
what you are doing, \fBDO NOT USE THIS TO FEED MAIL TO SENDMAIL!\fP.
.TP
.BI \-s\  script\-file\  \fR[not\ implemented\ yet]\fP
Specify additional scripts to be executed before the main script. Multiple
\fB\-s\fP arguments are allowed and the specified scripts are executed
sequentially in the order specified at the command
line.
.TP
.BI \-u\  user
Run the Sieve script for the given \fIuser\fP.
.TP
.B \-W
Enables write access to the source mailbox. This allows deleting the messages
from the source mailbox and changing the assigned IMAP flags and keywords. 
.TP
.BI \-x\  extensions
Set the available extensions. The parameter is a space\-separated list of the
active extensions. By prepending the extension identifiers with \fB+\fP or
\fB\-\fP, extensions can be included or excluded relative to the default set of
extensions. If no extensions have a \fB+\fP or \fB\-\fP prefix, only those
extensions that are explicitly listed will be enabled. Unknown extensions are
ignored and a warning is produced. By default, all supported extensions are
available, except for deprecated extensions or those that are still under
development.

For example \fB\-x\fP \(dq+imapflags \-enotify\(dq will enable the deprecated
imapflags extension along with all extensions that are available by default,
except for the enotify extension.

.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.I script\-file
Specifies the script to (compile and) execute.

Note that this tool looks for a pre\-compiled binary file with a \fI.svbin\fP
extension and with basename and path identical to the specified script. Use the
\fB\-C\fP option to disable this behavior by forcing the script to be compiled
into a new binary.
.TP
.I source\-mailbox
The name of the source mailbox.
.TP 
.I source\-action
Specifies what is done with messages in the source mailbox once processed by the
Sieve script. The \fIsource\-action\fP parameter accepts one of the following values:
.RS 7
.TP 
.BR keep\  (default)
Keep processed messages in source mailbox. When the filter decides to store the message
in the source mailbox, it is never duplicated there. However, in that case, the
IMAP flags of the original message can be modified by the Sieve interpreter. 
.TP 
.BI move\  mailbox
Move processed messages to the indicated \fImailbox\fP. This is for instance useful
to move messages to a Trash mailbox. 
.TP 
.B delete
Flag processed messages as \\DELETED.
.TP 
.B expunge
Expunge processed messages, meaning that these are removed irreversibly when the
tool finishes filtering.
.RE
.IP
The \fBsource\-action\fP is normally applied to all messages in the source mailbox,
but there are a few exceptions:
.RS 7
.IP \(bu
When the \fB\-W\fP option is not specified, the source mailbox is immutable and
the specified \fIsource\-action\fP has no effect.
.IP \(bu
When the \fB\-M\fR option (move mode) is active, all messages that were
successfully moved to another mailbox are expunged irrespective of the specified 
\fIsource\-action\fP. 
.IP \(bu
If the filter decides to keep the message in the source mailbox, it is left there
and not affected by the \fIsource\-action\fP.
.RE

.\"------------------------------------------------------------------------

.SH EXAMPLES

.TP
[...]

.\"------------------------------------------------------------------------
.SH "EXIT STATUS"
.B sieve\-filter
will exit with one of the following values:
.TP 4
.B 0
Sieve filter applied successfully. (EX_OK, EXIT_SUCCESS)
.TP
.B 1
Operation failed. This is returned for almost all failures.
(EXIT_FAILURE)
.TP
.B 64
Invalid parameter given. (EX_USAGE)
.\"------------------------------------------------------------------------
.SH FILES
.TP
.I @pkgsysconfdir@/dovecot.conf
Dovecot\(aqs main configuration file.
.TP
.I @pkgsysconfdir@/conf.d/90\-sieve.conf
Sieve interpreter settings (included from Dovecot\(aqs main configuration file)
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH "SEE ALSO"
.BR dovecot (1),
.BR dovecot\-lda (1),
.BR sieve\-dump (1),
.BR sieve\-test (1),
.BR sievec (1),
.BR pigeonhole (7)


src/dovecot/dovecot-279/dovecot-pigeonhole/src/sieve-tools/sieve-filter.c
/* Copyright (c) 2002-2012 Pigeonhole authors, see the included COPYING file
 */

#include "lib.h"
#include "lib-signals.h"
#include "ioloop.h"
#include "env-util.h"
#include "str.h"
#include "str-sanitize.h"
#include "ostream.h"
#include "array.h"
#include "mail-namespace.h"
#include "mail-storage.h"
#include "mail-search-build.h"
#include "imap-utf7.h"

#include "sieve.h"
#include "sieve-extensions.h"
#include "sieve-binary.h"

#include "sieve-tool.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <pwd.h>
#include <sysexits.h>

/*
 * Print help
 */

static void print_help(void)
{
	printf(
"Usage: sieve-filter [-c <config-file>] [-C] [-D] [-e] [-m <default-mailbox>]\n"
"                    [-P <plugin>] [-q <output-mailbox>] [-Q <mail-command>]\n"
"                    [-s <script-file>] [-u <user>] [-W] [-x <extensions>]\n"
"                    <script-file> <source-mailbox> <source-action>\n"
	);
}

enum sieve_filter_source_action {
	SIEVE_FILTER_SACT_KEEP,          /* Always keep messages in source folder */ 
	SIEVE_FILTER_SACT_MOVE,          /* Move discarded messages to Trash folder */      
	SIEVE_FILTER_SACT_DELETE,        /* Flag discarded messages as \DELETED */
	SIEVE_FILTER_SACT_EXPUNGE        /* Expunge discarded messages */
};

struct sieve_filter_data {
	enum sieve_filter_source_action source_action;
	struct mailbox *move_mailbox;

	struct sieve_script_env *senv;
	struct sieve_binary *main_sbin;
	struct sieve_error_handler *ehandler;

	unsigned int execute:1;
	unsigned int source_write:1;
	unsigned int move_mode:1;
};

struct sieve_filter_context {
	const struct sieve_filter_data *data;

	struct mailbox_transaction_context *move_trans;

	struct ostream *teststream;
};

static int filter_message
(struct sieve_filter_context *sfctx, struct mail *mail)
{
	struct sieve_error_handler *ehandler = sfctx->data->ehandler; 
	struct sieve_script_env *senv = sfctx->data->senv;
	struct sieve_exec_status estatus;
	struct sieve_binary *sbin;
	struct sieve_message_data msgdata;
	const char *recipient, *sender;
	bool execute = sfctx->data->execute;
	bool source_write = sfctx->data->source_write;
	bool move_mode = sfctx->data->move_mode;
	int ret;

	sieve_tool_get_envelope_data(mail, &recipient, &sender);

	/* Initialize execution status */
	memset(&estatus, 0, sizeof(estatus));
	senv->exec_status = &estatus;

	/* Collect necessary message data */
	memset(&msgdata, 0, sizeof(msgdata));
	msgdata.mail = mail;
	msgdata.return_path = sender;
	msgdata.orig_envelope_to = recipient;
	msgdata.final_envelope_to = recipient;
	msgdata.auth_user = senv->username;
	(void)mail_get_first_header(mail, "Message-ID", &msgdata.id);

	/* Single script */
	sbin = sfctx->data->main_sbin;

	/* Execute script */
	if ( execute ) {
		ret = sieve_execute(sbin, &msgdata, senv, ehandler, NULL);
	} else {
		ret = sieve_test
			(sbin, &msgdata, senv, ehandler, sfctx->teststream, NULL);
	}

	/* Handle message in source folder */
	if ( ret > 0 ) {
		struct mailbox *move_box = sfctx->data->move_mailbox;
		enum sieve_filter_source_action source_action =
			sfctx->data->source_action;

		if ( !source_write ) {
			/* Do nothing */

		} else if ( estatus.keep_original  ) {
			sieve_info(ehandler, NULL, "message kept in source mailbox");

		} else if ( move_mode && estatus.message_saved ) {
			sieve_info(ehandler, NULL,
				"message expunged from source mailbox upon successful move");

			if ( execute )
				mail_expunge(mail);

		} else {
			switch ( source_action ) {
			/* Leave it there */
			case SIEVE_FILTER_SACT_KEEP:
				sieve_info(ehandler, NULL, "message left in source mailbox");
				break;
			/* Move message to indicated folder */
			case SIEVE_FILTER_SACT_MOVE:			
				sieve_info(ehandler, NULL, 
					"message in source mailbox moved to mailbox '%s'",
					mailbox_get_name(move_box));

				if ( execute && move_box != NULL ) {
					struct mailbox_transaction_context *t = sfctx->move_trans;
					struct mail_save_context *save_ctx;
	
				    save_ctx = mailbox_save_alloc(t);

					if ( mailbox_copy(&save_ctx, mail) < 0 ) {
						enum mail_error error;
						const char *errstr;
	
						errstr = mail_storage_get_last_error
							(mailbox_get_storage(move_box), &error);

						sieve_error(ehandler, NULL,
							"failed to move message to mailbox %s: %s",
							mailbox_get_name(move_box), errstr);
						return -1;
				    }
			
					mail_expunge(mail);
				}
				break;
			/* Flag message as \DELETED */
			case SIEVE_FILTER_SACT_DELETE:					
				sieve_info(ehandler, NULL, "message flagged as deleted in source mailbox");
				if ( execute )
					mail_update_flags(mail, MODIFY_ADD, MAIL_DELETED);
				break;
			/* Expunge the message immediately */
			case SIEVE_FILTER_SACT_EXPUNGE:
				sieve_info(ehandler, NULL, "message expunged from source mailbox");
				if ( execute )
					mail_expunge(mail);
				break;
			/* Unknown */
			default:
				i_unreached();
				break;
			}
		}
	}

	return ret;
}

/* FIXME: introduce this into Dovecot */
static void mail_search_build_add_flags
(struct mail_search_args *args, enum mail_flags flags, bool not)
{
	struct mail_search_arg *arg;

	arg = p_new(args->pool, struct mail_search_arg, 1);
	arg->type = SEARCH_FLAGS;
	arg->value.flags = flags;
	arg->not = not;

	arg->next = args->args;
	args->args = arg;
}

static int filter_mailbox
(const struct sieve_filter_data *sfdata, struct mailbox *src_box)
{
	struct sieve_filter_context sfctx;
	struct mailbox *move_box = sfdata->move_mailbox;
	struct sieve_error_handler *ehandler = sfdata->ehandler;
	struct mail_search_args *search_args;
	struct mailbox_transaction_context *t;
	struct mail_search_context *search_ctx;
	struct mail *mail;
	int ret = 1;

	/* Sync source mailbox */

	if ( mailbox_sync(src_box, MAILBOX_SYNC_FLAG_FULL_READ) < 0 ) {
		sieve_error(ehandler, NULL, "failed to sync source mailbox");
		return -1;
	}

	/* Initialize */

	memset(&sfctx, 0, sizeof(sfctx));
	sfctx.data = sfdata;

	/* Create test stream */
	if ( !sfdata->execute )
		sfctx.teststream = o_stream_create_fd(1, 0, FALSE);

	/* Start move mailbox transaction */

	if ( move_box != NULL ) {
		sfctx.move_trans = mailbox_transaction_begin
			(move_box, MAILBOX_TRANSACTION_FLAG_EXTERNAL);
	}

	/* Search non-deleted messages in the source folder */

	search_args = mail_search_build_init();
	mail_search_build_add_flags(search_args, MAIL_DELETED, TRUE);

	t = mailbox_transaction_begin(src_box, 0);
	search_ctx = mailbox_search_init(t, search_args, NULL);
	mail_search_args_unref(&search_args);

	/* Iterate through all requested messages */

	mail = mail_alloc(t, 0, NULL);
	while ( ret > 0 && mailbox_search_next(search_ctx, mail) > 0 ) {
		const char *subject, *date;
		uoff_t size = 0;
					
		/* Request message size */

		if ( mail_get_virtual_size(mail, &size) < 0 ) {
			if ( mail->expunged )
				continue;
			
			sieve_error(ehandler, NULL, "failed to obtain message size");
			continue;
		}

		if ( mail_get_first_header(mail, "date", &date) <= 0 )
			date = "";
		if ( mail_get_first_header(mail, "subject", &subject) <= 0 ) 
			subject = "";
		
		sieve_info(ehandler, NULL,
			"filtering: [%s; %"PRIuUOFF_T" bytes] %s", date, size,
			str_sanitize(subject, 40));
	
		ret = filter_message(&sfctx, mail);
	}
	mail_free(&mail);
	
	/* Cleanup */

	if ( mailbox_search_deinit(&search_ctx) < 0 ) {
		ret = -1;
	}

	if ( sfctx.move_trans != NULL ) {
		if ( ret < 0 ) {
			mailbox_transaction_rollback(&sfctx.move_trans);
		} else {
			if ( mailbox_transaction_commit(&sfctx.move_trans) < 0 ) {
				ret = -1;
			}
		}
	}

	if ( ret < 0 ) {
		mailbox_transaction_rollback(&t);
	} else {
		if ( mailbox_transaction_commit(&t) < 0 ) {
			ret = -1;
		}
	}

	if ( sfctx.teststream != NULL )
		o_stream_destroy(&sfctx.teststream);

	if ( ret < 0 ) return ret;

	/* Sync mailbox */

	if ( sfdata->execute ) {
		if ( mailbox_sync(src_box, MAILBOX_SYNC_FLAG_FULL_WRITE) < 0 ) {
			sieve_error(ehandler, NULL, "failed to sync source mailbox");
			return -1;
		}
	}
	
	return ret;
}

static const char *mailbox_name_to_mutf7(const char *mailbox_utf8)
{
	string_t *str = t_str_new(128);

	if (imap_utf8_to_utf7(mailbox_utf8, str) < 0)
		return mailbox_utf8;
	else
		return str_c(str);
}

/*
 * Tool implementation
 */

int main(int argc, char **argv) 
{
	struct sieve_instance *svinst;
	ARRAY_TYPE (const_string) scriptfiles;
	const char *scriptfile,	*src_mailbox, *dst_mailbox, *move_mailbox;
	struct sieve_filter_data sfdata;
	enum sieve_filter_source_action source_action = SIEVE_FILTER_SACT_KEEP;
	struct mail_user *mail_user;
	struct sieve_binary *main_sbin;
	struct sieve_script_env scriptenv;
	struct sieve_error_handler *ehandler;
	bool force_compile, execute, source_write, move_mode;
	struct mail_namespace *ns;
	struct mailbox *src_box = NULL, *move_box = NULL;
	enum mailbox_flags open_flags =
		MAILBOX_FLAG_KEEP_RECENT | MAILBOX_FLAG_IGNORE_ACLS;
	enum mail_error error;
	int c;

	sieve_tool = sieve_tool_init("sieve-filter", &argc, &argv, 
		"m:s:x:P:u:q:Q:DCeWM", FALSE);

	t_array_init(&scriptfiles, 16);

	/* Parse arguments */
	scriptfile =  NULL;
	src_mailbox = dst_mailbox = move_mailbox = NULL;
	force_compile = execute = source_write = move_mode = FALSE;
	while ((c = sieve_tool_getopt(sieve_tool)) > 0) {
		switch (c) {
		case 'm':
			/* default mailbox (keep box) */
			dst_mailbox = optarg;
			break;
		case 's': 
			/* scriptfile executed before main script */
			{
				const char *file;			

				file = t_strdup(optarg);
				array_append(&scriptfiles, &file, 1);

				/* FIXME: */
				i_fatal_status(EX_USAGE, 
					"The -s argument is currently NOT IMPLEMENTED");
			}
			break;
		case 'q': 
			i_fatal_status(EX_USAGE, 
				"The -q argument is currently NOT IMPLEMENTED");
			break;
		case 'Q': 
			i_fatal_status(EX_USAGE, 
				"The -Q argument is currently NOT IMPLEMENTED");
			break;
		case 'e':
			/* execution mode */
			execute = TRUE;
			break;
		case 'C':
			/* force script compile */
			force_compile = TRUE;
			break;
		case 'M':
			/* move mode */
			move_mode = TRUE;
			break;
		case 'W':
			/* enable source mailbox write */
			source_write = TRUE;
			break;
		default:
			/* unrecognized option */
			print_help();
			i_fatal_status(EX_USAGE, "Unknown argument: %c", c);
			break;
		}
	}

	/* Script file argument */	
	if ( optind < argc ) {
		scriptfile = t_strdup(argv[optind++]);
	} else {
		print_help();
		i_fatal_status(EX_USAGE, "Missing <script-file> argument");
	}

	/* Source mailbox argument */
	if ( optind < argc ) {
		src_mailbox = t_strdup(argv[optind++]);
	} else {
		print_help();
		i_fatal_status(EX_USAGE, "Missing <source-mailbox> argument");
	}
		
	/* Source action argument */
	if ( optind < argc ) {
		const char *srcact = argv[optind++];

		if ( strcmp(srcact, "keep") == 0 ) {
			source_action = SIEVE_FILTER_SACT_KEEP;
		} else if ( strcmp(srcact, "move") == 0 ) {
			source_action = SIEVE_FILTER_SACT_MOVE;
			if ( optind < argc ) {
				move_mailbox = t_strdup(argv[optind++]);
			} else {
				print_help();
				i_fatal_status(EX_USAGE,
					"Invalid <source-action> argument: "
					"the `move' action requires mailbox argument");
			}
		} else if ( strcmp(srcact, "flag") == 0 ) {
			source_action = SIEVE_FILTER_SACT_DELETE;
		} else if ( strcmp(srcact, "expunge") == 0 ) {
			source_action = SIEVE_FILTER_SACT_EXPUNGE;
		} else {
			print_help();
			i_fatal_status(EX_USAGE, "Invalid <source-action> argument");
		}
	} 

	if ( optind != argc ) {
		print_help();
		i_fatal_status(EX_USAGE, "Unknown argument: %s", argv[optind]);
	}

	if ( dst_mailbox == NULL ) {
		dst_mailbox = src_mailbox;
	}

	/* Finish tool initialization */
	svinst = sieve_tool_init_finish(sieve_tool, TRUE);

        /* Enable debug extension */
        sieve_enable_debug_extension(svinst);

	/* Create error handler */
	ehandler = sieve_stderr_ehandler_create(svinst, 0);
	sieve_system_ehandler_set(ehandler);
	sieve_error_handler_accept_infolog(ehandler, TRUE);

	/* Compile main sieve script */
	if ( force_compile ) {
		main_sbin = sieve_tool_script_compile(svinst, scriptfile, NULL);
		if ( main_sbin != NULL )
			(void) sieve_save(main_sbin, NULL, TRUE, NULL);
	} else {
		main_sbin = sieve_tool_script_open(svinst, scriptfile);
	}

	/* Initialize mail user */
	mail_user = sieve_tool_get_mail_user(sieve_tool);

	/* Open the source mailbox */	

	src_mailbox = mailbox_name_to_mutf7(src_mailbox);
	ns = mail_namespace_find(mail_user->namespaces, &src_mailbox);
	if ( ns == NULL )
		i_fatal("Unknown namespace for source mailbox '%s'", src_mailbox);

	if ( !source_write || !execute ) 
		open_flags |= MAILBOX_FLAG_READONLY;
	
	src_box = mailbox_alloc(ns->list, src_mailbox, open_flags);
	if ( mailbox_open(src_box) < 0 ) {
		i_fatal("Couldn't open source mailbox '%s': %s", 
			src_mailbox, mail_storage_get_last_error(ns->storage, &error));
	}
	
	/* Open move box if necessary */

	if ( execute && source_action == SIEVE_FILTER_SACT_MOVE &&
		move_mailbox != NULL ) {
		move_mailbox = mailbox_name_to_mutf7(move_mailbox);
		ns = mail_namespace_find(mail_user->namespaces, &move_mailbox);
		if ( ns == NULL )
			i_fatal("Unknown namespace for mailbox '%s'", move_mailbox);

		move_box = mailbox_alloc(ns->list, move_mailbox, open_flags);
		if ( mailbox_open(move_box) < 0 ) {
			i_fatal("Couldn't open mailbox '%s': %s", 
				move_mailbox, mail_storage_get_last_error(ns->storage, &error));
		}

		if ( mailbox_backends_equal(src_box, move_box) ) {
			i_fatal("Source mailbox and mailbox for move action are identical.");
		}
	}

	/* Compose script environment */
	memset(&scriptenv, 0, sizeof(scriptenv));
	scriptenv.mailbox_autocreate = FALSE;
	scriptenv.default_mailbox = dst_mailbox;
	scriptenv.user = mail_user;
	scriptenv.username = sieve_tool_get_username(sieve_tool);
	scriptenv.hostname = "host.example.com";
	scriptenv.postmaster_address = "postmaster@example.com";
	scriptenv.smtp_open = NULL;
	scriptenv.smtp_close = NULL;

	/* Compose filter context */
	memset(&sfdata, 0, sizeof(sfdata));
	sfdata.senv = &scriptenv;
	sfdata.source_action = source_action;
	sfdata.move_mailbox = move_box;
	sfdata.main_sbin = main_sbin;
	sfdata.ehandler = ehandler;
	sfdata.execute = execute;
	sfdata.source_write = source_write;
	sfdata.move_mode = move_mode;

	/* Apply Sieve filter to all messages found */
	(void) filter_mailbox(&sfdata, src_box);
	
	/* Close the source mailbox */
	if ( src_box != NULL )
		mailbox_free(&src_box);

	/* Close the move mailbox */
	if ( move_box != NULL )
		mailbox_free(&move_box);

	/* Cleanup error handler */
	sieve_error_handler_unref(&ehandler);

	sieve_tool_deinit(&sieve_tool);

	return 0;
}
